### PR TITLE
[bugfix] custom SD: when ip out of order, reflect.deepEqual can not correctly identify whether there is a change

### PR DESCRIPTION
**Pull request by liguozhong**
[bugfix] custom SD: when ip out of order, reflect.deepEqual can not correctly identify whether there is a change

### CODE REVIEW

**Comment by liguozhong** [comment]
> It would be simpler to sort the targets alphabetically and keep `reflect.DeepEqual`.
 Thanks for your advice，According to your suggestion, I can fix this bug with just one line of code！This version of the code is already running in our production environment.

        sort.Strings(newTargets[:])

At the same time, my company's custom SD test case has also passed.

{
			title: "Disordered Ips in Alibaba's application management system",
			targetGroup: map[string][]*targetgroup.Group{
				"buy": {
					{
						Source: "alibaba",
						Targets: []model.LabelSet{
							{
								model.AddressLabel: "192.168.1.22",
							},
							{
								model.AddressLabel: "192.168.1.33",
							},
						},
						Labels: model.LabelSet{
							model.LabelName("__meta_test_label"): model.LabelValue("label_test_1"),
						},
					},
				},
				"cart": {
					{
						Source: "alibaba",
						Targets: []model.LabelSet{
							{
								model.AddressLabel: "192.168.1.55",
							},
							{
								model.AddressLabel: "192.168.1.44",
							},
						},
						Labels: model.LabelSet{
							model.LabelName("__meta_test_label"): model.LabelValue("label_test_1"),
						},
					},
				},
			},
			expectedCustomSD: map[string]*customSD{
				"buy:alibaba:21c0d97a1e27e6fe": {
					Targets: []string{
						"192.168.1.22",
						"192.168.1.33",
					},
					Labels: map[string]string{
						"__meta_test_label": "label_test_1",
					},
				},
				"cart:alibaba:1112e97a13b159fa": {
					Targets: []string{
						"192.168.1.44",
						"192.168.1.55",
					},
					Labels: map[string]string{
						"__meta_test_label": "label_test_1",
					},
				},
			},
		},

**Comment by liguozhong** [comment]
> It would be simpler to sort the targets alphabetically and keep `reflect.DeepEqual`.

Please help us to review the new commit again. thank you

File: documentation/examples/custom-sd/adapter/adapter.go [hunk 1]
```diff
@@ -18,16 +18,16 @@ import (
...
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/discovery"
 	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
```
Comment (simonpasquier) [inline documentation/examples/custom-sd/adapter/adapter.go]: We separate imports into 3 groups: stdlib, 3rd-party packages and local packages. These lines need to go back after the `"fmt"` line.

File: documentation/examples/custom-sd/adapter/adapter.go [hunk 2]
```diff
@@ -76,7 +76,7 @@ func generateTargetGroups(allTargetGroups map[string][]*targetgroup.Group) map[s
 					newTargets = append(newTargets, string(target))
 				}
 			}
-
+			sort.Strings(newTargets[:])
```
Comment (simonpasquier) [inline documentation/examples/custom-sd/adapter/adapter.go]: `sort.Strings(newTargets)` is enough

**Comment by liguozhong** [comment]
> The DCO check is failing.
I have modified the code as you said, please help review it again. thank you very much

File: documentation/examples/custom-sd/adapter/adapter.go (hunk 1 above)
Comment (simonpasquier) [reply documentation/examples/custom-sd/adapter/adapter.go]: you need an empty line after `"sort"`

**Comment by simonpasquier** [comment]
For the DCO:

```
You have 4 commits incorrectly signed off. To fix, head to your local branch and run:

git rebase HEAD~4 --signoff

Now your commits will have your sign off. Next run

git push --force-with-lease origin master
```

**Comment by simonpasquier** [comment]
You still need to sign the commits to comply with the DCO. See https://github.com/prometheus/prometheus/pull/5856/checks

**Comment by liguozhong** [comment]
> You still need to sign the commits to comply with the DCO. See https://github.com/prometheus/prometheus/pull/5856/checks

My git version 2.8.1 does not support the "git rebase HEAD~6 --signoff" command. After upgrading the git 2.23.0 version, this command can be passed.
image

File: documentation/examples/custom-sd/adapter/adapter.go (hunk 2 above)
Comment (liguozhong) [reply documentation/examples/custom-sd/adapter/adapter.go]: done

File: documentation/examples/custom-sd/adapter/adapter.go (hunk 1 above)
Comment (liguozhong) [reply documentation/examples/custom-sd/adapter/adapter.go]: done

File: documentation/examples/custom-sd/adapter/adapter_test.go [hunk 3]
```diff
@@ -151,6 +151,63 @@ func TestGenerateTargetGroups(t *testing.T) {
...
+				},
+			},
+			expectedCustomSD: map[string]*customSD{
+				"buy:alibaba:21c0d97a1e27e6fe": {
+					Targets: []string{
+						"192.168.1.22",
+						"192.168.1.33",
+					},
+					Labels: map[string]string{
+						"__meta_test_label": "label_test_2",
```
Comment (simonpasquier) [inline documentation/examples/custom-sd/adapter/adapter_test.go]: `label_test_1`

File: documentation/examples/custom-sd/adapter/adapter_test.go [hunk 4]
```diff
@@ -151,6 +151,63 @@ func TestGenerateTargetGroups(t *testing.T) {
...
+					},
+				},
+				"cart:alibaba:1112e97a13b159fa": {
+					Targets: []string{
+						"192.168.1.44",
+						"192.168.1.55",
+					},
+					Labels: map[string]string{
+						"__meta_test_label": "label_test_2",
```
Comment (simonpasquier) [inline documentation/examples/custom-sd/adapter/adapter_test.go]: `label_test_1`

**Comment by liguozhong** [comment]
> Glad you managed to fix the DCO issue! You're almost there...

done，Please help us to review the new commit again. Thank you!

image

**Pull request merged**